### PR TITLE
Domain Slippage Calculation

### PR DESCRIPTION
**Pull request by fleupold**
# Description
Follow up PR for #2661

This PR adds slippage tolerance to the AMM interactions when encoding their execution.

# Changes
- [x] Introduce slippage modules which takes a similar role as `SlippageContext` and `SlippageCalculator` in the old codebase
- [x] Expose auction prices to settlement (as those are required for slippage)
- [x] Reuse exsiting `in_eth` method on `Ether` and add an equivalent `from_eth` method for conversion in and out of native tokens.
- [x] Unit tests for slippage calculations

## How to test
Unit tests + CI on next PR, which makes new encoding the default

## Related Issues

Part of #2215

### CODE REVIEW

File: crates/driver/src/domain/competition/solution/slippage.rs [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+pub struct Interaction {
+    pub input: eth::Asset,
+    pub output: eth::Asset,
+}
+
+impl Parameters {
+    /// Apply the slippage parameters to the given liquidity increasing the in
+    /// amount by the appropriate slippage while keeping the out amount the
+    /// same.
+    pub fn apply_to(
```
Comment (fleupold) [inline crates/driver/src/domain/competition/solution/slippage.rs:35]: Old logic is here:

https://github.com/cowprotocol/services/blob/ab4e96efb6d8642d47e66ad54fc02c41688365d8/crates/solver/src/liquidity/slippage.rs#L44-L89

File: crates/driver/src/domain/competition/solution/slippage.rs [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        );
+        assert_eq!(exact_output.0.amount, interaction.output.amount);
+
+        // min cap
+        let min_cap = Parameters {
+            // 0.1 ETH (1 GNO)
+            min: Some(eth::U256::exp10(17)),
+            ..slippage.clone()
+        };
+        let (max_input, _) = min_cap.apply_to(&interaction).unwrap();
```
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/solution/slippage.rs:163]: we might as well also assert that the output amount is unchanged we the min and max cap.

File: crates/driver/src/domain/competition/auction.rs [hunk 3]
```diff
@@ -413,6 +413,22 @@ impl Price {
     pub fn in_eth(self, amount: eth::TokenAmount) -> eth::Ether {
         (amount.0 * self.0 .0 / Self::BASE).into()
     }
+
+    /// Convert an amount of ETH into a token amount using this price.
+    ///
+    /// Converting 1 ETH into GNO
+    ///
+    /// # Examples
+    /// ```
```
Comment (MartinquaXD) [inline crates/driver/src/domain/competition/auction.rs:422]: I think we have to set up the compilation of doc tests, right?
Without that these examples might get out of date very fast.

Comment (fleupold) [reply crates/driver/src/domain/competition/auction.rs:422]: https://github.com/cowprotocol/services/pull/2675

**Review by MartinquaXD — APPROVED** [review]

File: crates/driver/src/domain/competition/solution/encoding.rs [hunk 4]
```diff
@@ -0,0 +1,389 @@
...
+                        app_data: trade.order().app_data.0 .0.into(),
+                        fee_amount: eth::U256::zero(),
+                        flags: Flags {
+                            side: trade.order().side,
+                            partially_fillable: trade.order().partially_fillable,
+                            signing_scheme: trade.order().signature.scheme,
+                            sell_token_balance: trade.order().sell_token_balance,
+                            buy_token_balance: trade.order().buy_token_balance,
+                        },
+                        executed_amount: trade.executed().0,
```
Comment (Mateo-mro) [inline crates/driver/src/domain/competition/solution/encoding.rs:136]: nit:
```suggestion
                        executed_amount: trade.executed().into(),
```

File: crates/driver/src/domain/competition/solution/encoding.rs [hunk 5]
```diff
@@ -0,0 +1,389 @@
...
+                )
+            }
+            super::Trade::Jit(trade) => {
+                (
+                    Price {
+                        // Jit orders are matched at limit price, so the sell token is worth
+                        // buy.amount and vice versa
+                        sell_token: trade.order().sell.token.0.into(),
+                        sell_price: trade.order().buy.amount.into(),
+                        buy_token: trade.order().buy.token.0.into(),
```
Comment (Mateo-mro) [inline crates/driver/src/domain/competition/solution/encoding.rs:117]: nit:
```suggestion
                        buy_token: trade.order().buy.token.into(),
```

**Review by Mateo-mro — APPROVED** [review]

**Review by squadgazzz — APPROVED** [review]

**Pull request merged**